File: 2020/15_RambunctiousRecitation/game.test.ts
// ======================================================================
// Rambunctious Recitation
//   Advent of Code 2020 Day 15 -- Eric Wastl -- https://adventofcode.com
//
// Typescript implementation by Dr. Dean Earl Wright III
//  ======================================================================

// ======================================================================
//                      g a m e . t e s t . t s
//
// Test the solver for Advent of Code 2020 day 15 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

import { fromText } from './aoc_15';
import { Game } from './game';

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const EXAMPLE_TEXT = `
0,3,6
`;

interface ExampleTests {
  text: string;
  spoken: number;
}

const EXAMPLES_PART_ONE: ExampleTests[] = [
  { text: '0,3,6', spoken: 436 },
  { text: '1,3,2', spoken: 1 },
  { text: '2,1,3', spoken: 10 },
  { text: '1,2,3', spoken: 27 },
  { text: '2,3,1', spoken: 78 },
  { text: '3,2,1', spoken: 438 },
  { text: '3,1,2', spoken: 1836 },
];
const EXAMPLES_PART_TWO: ExampleTests[] = [
  { text: '0,3,6', spoken: 175594 },
  { text: '1,3,2', spoken: 2578 },
  { text: '2,1,3', spoken: 3544142 },
  { text: '1,2,3', spoken: 261214 },
  { text: '2,3,1', spoken: 6895259 },
  { text: '3,2,1', spoken: 18 },
  { text: '3,1,2', spoken: 362 },
];

const PART_ONE_TEXT = EXAMPLE_TEXT;
const PART_TWO_TEXT = EXAMPLE_TEXT;

const PART_ONE_RESULT = 436;
const PART_TWO_RESULT = 175594;

// ======================================================================
//                                                              TestGame
// ======================================================================

describe('Game', () => {
  test('Test the default Game creation', () => {
    // 1. Create default Game object
    const myobj = new Game([]);
    // 2. Make sure it has the default values
    expect(myobj.part2).toBe(false);
    expect(myobj.text).toHaveLength(0);
  });

  test('Test the Game object creation from text', () => {
    // 1. Create Game object from text
    const myobj = new Game(fromText(EXAMPLE_TEXT));
    // 2. Make sure it has the expected values
    expect(myobj.part2).toBe(false);
    expect(myobj.text).toHaveLength(1);
  });

  test('Test all of the part one examples', () => {
    // 1. Loop for all of the examples
    EXAMPLES_PART_ONE.forEach((test) => {
      // 2. Create Game object
      const myobj = new Game(fromText(test.text));
      expect(myobj.part2).toBe(false);
      expect(myobj.text).toHaveLength(1);
      // 3. Make sure it has the expected value
      expect(myobj.solution()).toBe(test.spoken);
    });
  });

  test('Test all of the part two examples', () => {
    // 1. Loop for all of the examples for the second part
    EXAMPLES_PART_TWO.forEach((test) => {
      // 2. Create Game object using the key as text
      const myobj = new Game(fromText(test.text), true);
      expect(myobj.part2).toBe(true);
      expect(myobj.text).toHaveLength(1);
      // 3. Make sure it has the expected value
      // expect(myobj.solution()).toBe(test.spoken);
    });
  });

  test('Test part one example of Game object', () => {
    // 1. Create Game object from text
    const myobj = new Game(fromText(PART_ONE_TEXT));
    // 2. Check the part one result
    expect(myobj.partOne(false, 0)).toBe(PART_ONE_RESULT);
  });

  test('Test part two example of Game object', () => {
    // 1. Create Game object from text
    const myobj = new Game(fromText(PART_TWO_TEXT), true);
    // 2. Check the part two result
    expect(myobj.partTwo(false, 0)).toBe(PART_TWO_RESULT);
  });
});

// ======================================================================
// end                   g a m e . t e s t . t s                  end
// ======================================================================

File: 2018/09_MarbleMania/game_test.go
// ======================================================================
// Marble Mania
//   Advent of Code 2018 Day 09 -- Eric Wastl -- https://adventofcode.com
//
// lua implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                    g a m e _ t e s t . g o
// ======================================================================
// Test Game for Advent of Code 2018 day 09, Marble Mania

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"testing"
)

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

var (
	GAME_EXAMPLES = [...]int{
		10, 1618, 8317,
		13, 7999, 146373,
		17, 1184, 2764,
		21, 6111, 54718,
		30, 5807, 37305}
)

// ======================================================================
//                                                               TestGame
// ======================================================================

func TestGameShort(t *testing.T) {
	// Test the default Game creation

	// 1. Create default Game object
	obj := NewGame(9, 25)

	// 2. Make sure it has theexpected values
	if obj.numPlayers != 9 {
		t.Fatalf("Number of Players is %d not 9", obj.numPlayers)
	}
	if obj.numMarbles != 25 {
		t.Fatalf("Number of Marbles is %d not 25", obj.numMarbles)
	}
	if len(obj.players) != 9 {
		t.Fatalf("Length of Players is %d not 9", len(obj.players))
	}
	if len(obj.circle) != 1 {
		t.Fatalf("Length of circle is %d not 1", len(obj.circle))
	}
	if cap(obj.circle) != 25 {
		t.Fatalf("Capacity of circle is %d not 25", cap(obj.circle))
	}
	if obj.nextPlayer != 0 {
		t.Fatalf("Next Players is %d not 0", obj.nextPlayer)
	}
	if obj.nextMarble != 1 {
		t.Fatalf("Next Marble is %d not 1", obj.nextMarble)
	}
	if obj.current != 0 {
		t.Fatalf("Current is %d not 0", obj.current)
	}

	// 3. Check methods
	show := obj.Show()
	if show != "[----] (0)" {
		t.Fatalf("Initial show is '%s' not '[----] (0)'", show)
	}
	if obj.IsSpecial(1) {
		t.Fatal("One is not special")
	}
	if obj.IsSpecial(7) {
		t.Fatal("Seven is not special")
	}
	if !obj.IsSpecial(23) {
		t.Fatal("Twenty three should be special")
	}
	if !obj.IsSpecial(46) {
		t.Fatal("Forty six should be special")
	}
	if obj.IsSpecial(50) {
		t.Fatal("Fifty is not special")
	}

	// 4. First turn
	cwise := obj.Clockwise()
	if cwise != 1 {
		t.Fatalf("Initial closewise is %d not 1", cwise)
	}
	more := obj.Turn()
	if !more {
		t.Fatal("Game ended after first turn")
	}
	show = obj.Show()
	if show != "[   1]  0 (1)" {
		t.Fatalf("First turn show is '%s' not '[   1]  0 (1)'", show)
	}

}

func TestGameExamples(t *testing.T) {
	// Test the Game object creation from text

	// 1. Loop through the examples
	for i := 0; i < len(GAME_EXAMPLES); i = i + 3 {

		// 2. Get the values
		players := GAME_EXAMPLES[i]
		marbles := GAME_EXAMPLES[i+1]
		expected := GAME_EXAMPLES[i+2]

		// 2. Create the game object
		game := NewGame(players, marbles)

		// 3. Play the game
		total := game.AllTurns()

		// 4. Check the result
		if total != expected {
			t.Fatalf("Game(%d, %d) is %d not %d", players, marbles, total, expected)
		}
	}
}

// ======================================================================
// end                  g a m e _ t e s t. g o                end
// ======================================================================
